File: go/secret-handshake/secret_handshake.go
// Package secret implements a secret handshake for a code
package secret

import (
	"fmt"
)

const (
	testVersion = 1
	runeOne     = 49
)

var ops = map[int]string{
	0: "wink",
	1: "double blink",
	2: "close your eyes",
	3: "jump",
}

// Handshake shows the operations for the secret handshake
func Handshake(code uint) []string {
	var handshake []string

	b := fmt.Sprintf("%b", code)
	b = reverse(b)

	fmt.Println(code | 0x02)

	for index, r := range b {
		if index == 4 {
			handshake = reverseStrArray(handshake)
			return handshake // no need to evaluate greater than 32
		} else if index < 4 {
			if r == runeOne {
				handshake = append(handshake, ops[index])
			}
		}
	}

	return handshake
}

func reverse(s string) string {
	var reverse string
	for i := len(s) - 1; i >= 0; i-- {
		reverse += string(s[i])
	}
	return reverse
}

func reverseStrArray(s []string) []string {
	for i := len(s)/2 - 1; i >= 0; i-- {
		opp := len(s) - 1 - i
		s[i], s[opp] = s[opp], s[i]
	}
	return s
}

File: elixir/secret-handshake/lib/secret_handshake.ex
defmodule SecretHandshake do
  @actions ["jump", "close your eyes", "double blink", "wink"]

  @doc """
  Determine the actions of a secret handshake based on the binary
  representation of the given `code`.

  If the following bits are set, include the corresponding action in your list
  of commands, in order from lowest to highest.

  1 = wink
  10 = double blink
  100 = close your eyes
  1000 = jump

  10000 = Reverse the order of the operations in the secret handshake
  """
  @spec commands(code :: integer) :: list(String.t())

  def commands(code) when code > 31 or code < 1 or code == 16 do
    []
  end

  def commands(code) when code > 16 and code < 32 do
    do_commands(code - 16)
  end

  def commands(code) do
    do_commands(code)
    |> Enum.reverse()
  end

  def do_commands(code) do
    code
    |> Integer.to_string(2)
    |> String.pad_leading(4, "0")
    |> String.codepoints()
    |> Enum.zip(@actions)
    |> Enum.filter(fn x -> elem(x, 0) == "1" end)
    |> Enum.map(fn x -> elem(x, 1) end)
  end
end
